torch/csrc/jit/interpreter.cpp

```diff
@@ -0,0 +1,288 @@
+#include "interpreter.h"
+#include "torch/csrc/jit/ir.h"
+#include "torch/csrc/jit/generated/aten_dispatch.h"
+#ifdef WITH_CUDA
+#include "torch/csrc/jit/fusion_compiler.h"
+#endif
+
+namespace torch { namespace jit {
+
+using InputList = const std::vector<at::Tensor> &;
+using OutputList = std::vector<at::Tensor>&;
+using Callback = std::function<void(InputList, OutputList)>;
+// Returns a function implementing functionality of a given node,
+// or nullptr if it's a no-op for autograd.
+Callback getCallback(Node *node) {
+  IR_IFM(node, PythonOp)
+    throw NotImplementedException();
+  IR_ELSEIFM(CppOp)
+    throw NotImplementedException();
+  IR_ELSEIF(Select)
+    barf("getCallback() on select?");
+  IR_ELSEIF(FusionGroup)
+#ifdef WITH_CUDA
+    auto fusion_fn = sharedFusionCompiler().getOrCompile(*value->g(kSubgraph));
+    return [fusion_fn](InputList inputs, OutputList outputs) {
+      fusion_fn->launch(inputs, outputs);
+    };
+#else
+    throw std::runtime_error("don't know how to execute FusionGroups without CUDA");
+#endif
+  IR_ELSEIF(Constant)
+    auto t = value->t(kvalue);
+    return [t](InputList inputs, OutputList outputs) {
+      outputs.push_back(t);
```

Comment:
Err... do we allow inplace operations in the new interpreter? If so, wouldn't we get in trouble if we did an inplace op on this constant? I remember thinking we had this hazard in the old code too, but we had a clone to prevent this from happening:

```
struct WrapConstant : public Function {
  WrapConstant(at::Tensor value)
    : value(std::move(value)) {
    is_executable = true;
    num_inputs = 1;
  }

  virtual variable_list apply(const variable_list& inputs) {
    if (inputs.size() != 1 || inputs[0].defined())
      throw std::logic_error("WrapConstant nodes should only receive a single NULL input");
    AutoGPU guard(value);
    return {make_variable(value.clone())};
  }

  at::Tensor value;
};
```

Reply:
I am assuming the Graph doesn't have any inplace operators. What we can do later is optimize the instruction generation to re-use inputs tensors when an inplace variant exists and it is safe to do (no more uses, not ruining a constant). Copying constants can potentially add huge overhead, so it makes sense to stick with working with the fast path.